pkg/controller/apis/config/v1alpha1/defaults.go

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"time"
+
+	apimachineryconfigv1alpha1 "k8s.io/apimachinery/pkg/apis/config/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kruntime "k8s.io/apimachinery/pkg/runtime"
+	apiserverconfigv1alpha1 "k8s.io/apiserver/pkg/apis/config/v1alpha1"
+	kubectrlmgrconfigv1alpha1 "k8s.io/kube-controller-manager/config/v1alpha1"
+	utilpointer "k8s.io/utils/pointer"
+)
+
+func addDefaultingFuncs(scheme *kruntime.Scheme) error {
+	return RegisterDefaults(scheme)
+}
+
+func SetDefaults_KubeControllerManagerConfiguration(obj *kubectrlmgrconfigv1alpha1.KubeControllerManagerConfiguration) {
+	zero := metav1.Duration{}
+
+	if obj.AttachDetachController.ReconcilerSyncLoopPeriod == zero {
+		obj.AttachDetachController.ReconcilerSyncLoopPeriod = metav1.Duration{Duration: 60 * time.Second}
+	}
+	if obj.DaemonSetController.ConcurrentDaemonSetSyncs == 0 {
+		obj.DaemonSetController.ConcurrentDaemonSetSyncs = 2
+	}
+	if obj.DeprecatedController.RegisterRetryCount == 0 {
+		obj.DeprecatedController.RegisterRetryCount = 10
+	}
+	if obj.EndpointController.ConcurrentEndpointSyncs == 0 {
+		obj.EndpointController.ConcurrentEndpointSyncs = 5
+	}
+	if obj.NodeIPAMController.NodeCIDRMaskSize == 0 {
+		obj.NodeIPAMController.NodeCIDRMaskSize = 24
+	}
+	if obj.PersistentVolumeBinderController.PVClaimBinderSyncPeriod == zero {
+		obj.PersistentVolumeBinderController.PVClaimBinderSyncPeriod = metav1.Duration{Duration: 15 * time.Second}
+	}
+	if obj.PodGCController.TerminatedPodGCThreshold == 0 {
+		obj.PodGCController.TerminatedPodGCThreshold = 12500
+	}
+	if obj.ReplicationController.ConcurrentRCSyncs == 0 {
+		obj.ReplicationController.ConcurrentRCSyncs = 5
+	}
+	if obj.ReplicaSetController.ConcurrentRSSyncs == 0 {
+		obj.ReplicaSetController.ConcurrentRSSyncs = 5
+	}
+	if obj.SAController.ConcurrentSATokenSyncs == 0 {
+		obj.SAController.ConcurrentSATokenSyncs = 5
+	}
+
+	if obj.Generic.ClientConnection.QPS == 0.0 {
+		obj.Generic.ClientConnection.QPS = 20.0
+	}
+	if obj.Generic.ClientConnection.Burst == 0 {
+		obj.Generic.ClientConnection.Burst = 30
+	}
+}
+
+func SetDefaults_GenericControllerManagerConfiguration(obj *kubectrlmgrconfigv1alpha1.GenericControllerManagerConfiguration) {
```

Comment:
If you have time, this should be written as `RecommendedDefaultGenericControllerManagerConfiguration` like e.g. `RecommendedDefaultClientConnectionConfiguration`, and invoked manually by `SetDefaults_KubeControllerManagerConfiguration` and `SetDefaults_CloudControllerManagerConfiguration` respectively